src/dmd/e2ir.d

```diff
@@ -1363,6 +1363,32 @@ elem *toElem(Expression e, IRState *irs)
                 FuncDeclaration fd = te.var.toParent2().isFuncDeclaration();
                 assert(fd);
                 ethis = getEthis(te.loc, irs, fd);
+
+                /* Bugzilla 15984: If te is in interface contracts, ethis is a
+                 * reference to concrete class instance. Offset it in here
+                 * to get correct interface type reference.
+                 */
+                if (fd.ident == Id.ensure || fd.ident == Id.require)
+                {
+                    if (auto parentfd = fd.parent.isFuncDeclaration())
+                    {
+                        if (auto iface = parentfd.parent.isInterfaceDeclaration())
+                        {
+                            Type thistype = irs.getFunc().vthis.type;
```

Comment:
@FeepingCreature In particular I'm not certain about the provenance of `irs.getFunc()` (c.f. [`gIR->func()`](https://github.com/ldc-developers/ldc/blob/9ebf72da2bdc4c8d4fd76a76b3fb35951dd9a906/gen/tocall.cpp#L744) and `fd` (c.f.[`dfnval->func`](https://github.com/ldc-developers/ldc/blob/9ebf72da2bdc4c8d4fd76a76b3fb35951dd9a906/gen/tocall.cpp#L732)
everything else I'm pretty sure is correct.